src/core/coap/coap_server.hpp

```diff
@@ -94,6 +96,193 @@ class Resource : public otCoapResource
 };
 
 /**
+ * This class implements metadata required for caching CoAP responses.
+ *
+ */
+class EnqueuedResponseHeader
+{
+public:
+    /**
+     * Default constructor creating empty object.
+     *
+     */
+    EnqueuedResponseHeader(void): mDequeueTime(0), mMessageInfo() {}
+
+    /**
+     * Constructor creating object with valid dequeue time and message info.
+     *
+     * @param[in]  aMessageInfo  The message info containing source endpoint identification.
+     *
+     */
+    EnqueuedResponseHeader(const Ip6::MessageInfo &aMessageInfo):
+        mDequeueTime(Timer::GetNow() + Timer::SecToMsec(kExchangeLifetime)),
+        mMessageInfo(aMessageInfo) {}
+
+    /**
+     * This method append metadata to the message.
+     *
+     * @param[in]  aMessage  A reference to the message.
+     *
+     * @retval kThreadError_None    Successfully appended the bytes.
+     * @retval kThreadError_NoBufs  Insufficient available buffers to grow the message.
+     */
+    ThreadError AppendTo(Message &aMessage) const { return aMessage.Append(this, sizeof(*this)); }
+
+    /**
+     * This method reads request data from the message.
+     *
+     * @param[in]  aMessage  A reference to the message.
+     *
+     * @returns The number of bytes read.
+     *
+     */
+    uint16_t ReadFrom(const Message &aMessage) {
+        return aMessage.Read(aMessage.GetLength() - sizeof(*this), sizeof(*this), this);
+    }
+
+    /**
+     * This method removes metadata from the message.
+     *
+     * @param[in]  aMessage  A reference to the message.
+     *
+     * @retval kThreadError_None    Successfully removed the bytes.
+     * @retval kThreadError_NoBufs  Insufficient available buffers to place the message.
```

Comment:
Called function returns `_NoBufs`. I followed the code and it seems impossible in this case. I'll assert it and change function type to void.